plugins/reports/counter/classes/CounterReport.inc.php

```diff
@@ -0,0 +1,301 @@
+ *
+ * @brief A COUNTER report, base class
+ */
+require_once(dirname(dirname(__FILE__)).'/classes/COUNTER/COUNTER.php');
+
+define('COUNTER_EXCEPTION_WARNING', 0);
+define('COUNTER_EXCEPTION_ERROR', 1);
+define('COUNTER_EXCEPTION_PARTIAL_DATA', 4);
+define('COUNTER_EXCEPTION_NO_DATA', 8);
+define('COUNTER_EXCEPTION_BAD_COLUMNS', 16);
+define('COUNTER_EXCEPTION_BAD_FILTERS', 32);
+define('COUNTER_EXCEPTION_BAD_ORDERBY', 64);
+define('COUNTER_EXCEPTION_BAD_RANGE', 128);
+define('COUNTER_EXCEPTION_INTERNAL', 256);
+
+define('COUNTER_CLASS_PREFIX', 'CounterReport');
+
+// COUNTER as of yet is not internationalized and requires English constants
+define('COUNTER_LITERAL_ARTICLE', 'Article');
+define('COUNTER_LITERAL_JOURNAL', 'Journal');
+define('COUNTER_LITERAL_PROPRIETARY', 'Proprietary');
+
+class CounterReport {
+
+	var $parentPluginCategory;
+	var $parentPluginName;
+	
+	var $_release;
+	
+	var $_errors;
+	
+	/**
+	 * Constructor
+	 * @param string $parentPluginCategory
+	 * @param string $parentPluginName
+	 */
+	function CounterReport($parentPluginCategory, $parentPluginName) {
+		$this->parentPluginCategory = $parentPluginCategory;
+		$this->parentPluginName = $parentPluginName;
+		$this->_release = $this->getParentPlugin()->getCurrentRelease();
+	}
+	
+
+	/**
+	 * Get the parent plugin
+	 * @return object
+	 */
+	function &getParentPlugin() {
+		$plugin =& PluginRegistry::getPlugin($this->parentPluginCategory, $this->parentPluginName);
+		return $plugin;
+	}
+
+	/**
+	 * Get the COUNTER Release
+	 * @return $string
+	 */
+	function getRelease() {
+		return $this->_release;
+	}
+
+	/**
+	 * Get the report code
+	 * @return $string
+	 */
+	function getCode() {
+		return substr(get_class($this), strlen(COUNTER_CLASS_PREFIX));
+	}
+
+	/**
+	 * Get the COUNTER metric type for an Statistics file type
+	 * @param $filetype string
+	 * @return string
+	 */
+	function getKeyForFiletype($filetype) {
+		switch ($filetype) {
+			case STATISTICS_FILE_TYPE_HTML:
+				$metricTypeKey = 'ft_html';
+				break;
+			case STATISTICS_FILE_TYPE_PDF:
+				$metricTypeKey = 'ft_pdf';
+				break;
+			case STATISTICS_FILE_TYPE_OTHER:
+			default:
+				$metricTypeKey = 'other';
+		}
+		return $metricTypeKey;
+	}
+
+	/**
+	 * Abstract method must be implemented in the child class
+	 * Get the report title
+	 * @return $string
+	 */
+	function getTitle() {
+		assert(false);
+	}
+
+	/*
+	 * Convert an OJS metrics request to COUNTER ReportItems
+	 * Abstract method must be implemented by subclass
+	 * @param $columns string|array column (aggregation level) selection
+	 * @param $filters array report-level filter selection
+	 * @param $orderBy array order criteria
+	 * @param $range null|DBResultRange paging specification
+	 * @see ReportPlugin::getMetrics for more details on parameters
+	 * @return array COUNTER\ReportItem array
+	 */
+	function getReportItems($columns = array(), $filters = array(), $orderBy = array(), $range = null) {
+		assert(false);
```

Comment:
This PHP4/5 hybrid was the result of this pkp-dev conversation:

From: Graham, Clinton T 
Sent: Monday, January 05, 2015 2:36 PM
To: pkp-dev
Subject: Exceptions and OJS 2.x codebase

> I need a way to pass error conditions from one plugin (reports/counter) to another (generic/sushiLite). The error conditions should include a classification scheme and a human readable message.
 
> The clear way to go would be PHP5 Exceptions.  We incorporate Exceptions in the 2.4.x codebase via multiple third party libraries (ZendFramework, ADODB, htmlpurifier, lessphp, pqp, swordappv2, tinymce’s jbimages) and in our own code (tests, tools, dataverse plugin, pln plugin, sword plugin).
 
> Given the existing codebase, I’m not feeling very reluctant to make the COUNTER and SUSHI-Lite plugins PHP5 dependent as well.  I’m throwing this note out there to solicit any concerns with that approach.

From: Alec Smecher
Sent: Monday, January 05, 2015 2:44 PM
To: Graham, Clinton T; pkp-dev
Subject: Re: Exceptions and OJS 2.x codebase

> Would this just be in plugins? If so, please go for it -- and if you're giving some thought to exception handling in general, maybe we should put you on the spot during one of the tech committee calls to talk about establishing some best practices.

> Gotcha: watch out for the overridden error handler (set_error_handler); that's currently present to prevent PHP parse errors in situations where PHP4 and PHP5 had vastly different reference behavior.

From: Graham, Clinton T
Sent: Monday, January 05, 2015 2:55 PM
To: Alec Smecher; pkp-dev
Subject: RE: Exceptions and OJS 2.x codebase

>It would be just in the plugins.  I’m not convinced yet that I’m even going to throw the Exception from one plugin to the other; I may just instanciate and return an array of Exceptions via a getErrors() method, which would bypass the set_error_handler entirely.